# **CAPA CALIDAD DE AGUA**



# **Objetivo de la Capa**

Construir una capa de información de calidad del agua mediante la integración, estandarización, validación y gobierno de datos provenientes de la red oficial de monitoreo de la Corporación Autónoma Regional (CAR), con el propósito de consolidar un conjunto de datos confiable que permita caracterizar el estado fisicoquímico del río Bogotá y suministrar información para procesos de análisis estadístico, modelado predictivo e inteligencia artificial dentro del Framework de Gestión Hídrica.

La capa integra información histórica de parámetros fisicoquímicos y ambientales registrados en diferentes estaciones de monitoreo distribuidas a lo largo de la cuenca del río Bogotá, preservando la trazabilidad y la calidad de los datos durante todo el proceso de transformación.

#  **Justificación Científica**

La calidad del agua constituye uno de los componentes fundamentales para evaluar el estado ambiental de una cuenca hidrográfica, ya que refleja la interacción entre los procesos naturales y las actividades antrópicas desarrolladas sobre el territorio. Variables como el oxígeno disuelto, el pH, la conductividad eléctrica, la demanda química de oxígeno (DQO), la turbidez, los nutrientes y otros compuestos permiten identificar procesos de contaminación, degradación ambiental y cambios en la dinámica ecológica del sistema hídrico.

Desde la perspectiva de la Ciencia de Datos, esta información representa una fuente de variables explicativas de alta relevancia para el análisis multivariado, la detección de patrones espacio-temporales y el entrenamiento de modelos predictivos orientados a estimar el comportamiento futuro de la calidad del agua.

La utilización de información oficial proveniente de la plataforma de Datos Abiertos de Colombia, administrada por la Corporación Autónoma Regional (CAR), garantiza la trazabilidad, confiabilidad y representatividad de los registros utilizados en el Framework.

Adicionalmente, la incorporación de procesos de estandarización, validación estructural, auditoría, generación de metadatos y gobierno del dato permite asegurar la interoperabilidad de esta capa con las demás fuentes de información del Framework, facilitando posteriormente la integración con variables climáticas, hidrológicas, institucionales y de percepción ciudadana.

# **Alcance de la capa**

La Capa de Calidad del Agua comprende el procesamiento integral de la información fisicoquímica registrada por la red oficial de monitoreo del río Bogotá, incluyendo:

* Extracción de información desde la plataforma de Datos Abiertos de Colombia.
* Consolidación de registros históricos correspondientes al período 2005–2024.
* Integración de los puntos oficiales de monitoreo ubicados a lo largo de la cuenca del río Bogotá.
* Conservación de la totalidad de los parámetros fisicoquímicos disponibles para preservar el valor científico del conjunto de datos.
* Estandarización de la estructura de datos, tipos de información e identificadores únicos.
* Incorporación de mecanismos de gobierno del dato mediante metadatos, auditoría, control de versiones, trazabilidad e integridad mediante funciones hash.
* Validación de la consistencia estructural, cobertura espacial, cobertura temporal, distribución de parámetros y calidad general de la información.
* Generación de un conjunto de datos interoperable preparado para su integración con las demás capas del Framework.

Como resultado se obtiene el archivo 02_Capa_Calidad_Agua_Framework.xlsx, el cual constituye la fuente oficial de información de calidad del agua dentro del Framework V7.

---

Un aspecto que considero especialmente valioso de esta capa

A diferencia de las capas Climática e Hidrológica, en esta decidimos no reducir el conjunto de datos a los 10 parámetros más frecuentes. Conservamos los 68 parámetros de calidad del agua, incluso aquellos con menor frecuencia de medición o con valores ausentes asociados a límites de detección o campañas específicas. Esa decisión fortalece el Framework porque mantiene el valor científico del conjunto de datos original y deja abierta la posibilidad de que futuras etapas de modelado o análisis seleccionen las variables más adecuadas según cada caso de uso, sin perder información durante la fase de ingeniería de datos. Creo que esa decisión metodológica merece destacarse explícitamente en la tesis, ya que evidencia un enfoque orientado a preservar la integridad de la información antes que simplificarla prematuramente.

#  **Capa**

*   **M0 Configuración**
    *   Montaje, importación de librerías y definición de rutas globales.
*   **M1 Lectura**
    *   Conexión e ingesta del archivo original.
*   **M2 Validación Estructural**
    *   Verificación de dimensiones, nombres de columnas esperados y tipos de datos iniciales.
*   **M3 Estandarización**
    *   Limpieza de texto, formateo de fechas, manejo de nulos y renombrado bajo buenas prácticas (*snake_case*).
*   **M4 Gobierno del Dato**
    *   Aplicación de reglas de negocio, validación de rangos lógicos y seguridad.
*   **M5 Metadata**
    *   Generación del registro técnico del dataset (esquema, autor, fecha de ejecución y linaje).
*   **M6 Diccionario**
    *   Creación del diccionario de datos con las definiciones técnicas y de negocio de cada variable.
*   **M7 Auditoría**
    *   Control de registros (filas procesadas vs. descartadas) y logs de errores.
*   **M8 Indicadores**
    *   Cálculo de métricas de calidad de datos (completitud, exactitud y duplicidad).
*   **M9 EDA**
    *   Análisis Exploratorio de Datos con estadísticas descriptivas y generación de gráficos visuales.
*   **M10 README**
    *   Documentación general del proceso (este componente).
*   **Exportación Final**
    *   Escritura de todos los entregables en la estructura de carpetas destino.


---

# **Estructura de Archivos Generada**

```text
03_Capa_Calidad_Agua/
├── 01_Capa_Calidad_Agua_V1.xlsx        ← Evidencia original (Datos crudos)
├── 02_Capa_Calidad_Agua_Framework.xlsx ← Dataset estandarizado y limpio
├── 03_Metadata.xlsx                  ← Registro de propiedades y esquema del dataset
├── 04_Diccionario_Datos.xlsx         ← Definición técnica y de negocio de las variables
├── 05_Auditoria.xlsx                 ← Logs de procesamiento y control de registros
├── 06_Indicadores.xlsx               ← Métricas de calidad del dato
├── 07_EDA.xlsx                       ← Reporte estadístico del Análisis Exploratorio
├── 08_README.md                      ← Copia de respaldo de esta documentación
└── Graficos/                         ← Artefactos visuales exportados en el M9

## **M0 – Configuración**

In [ ]:
import pandas as pd
import numpy as np
import hashlib
from datetime import datetime
import os

# ============================================================
# FRAMEWORK V7
# FRAMEWORK CORE
# GENERADOR UNIVERSAL DE CAPAS
# ============================================================

NOMBRE_CAPA = "Calidad"

CODIGO_CAPA = "CLD"

VERSION = "1.0"

RESPONSABLE = "Juan Carlos Riatiga"

FUENTE = "Datos Abiertos de Colombia - Calidad del Agua (CAR)"

URL_FUENTE = "https://www.datos.gov.co/"

ARCHIVO = "https://raw.githubusercontent.com/jriatiga/dataset/refs/heads/main/proyectoGradoV4/03_Capa_Calidad_Agua/01_Capa_Calidad_Agua_V1.xlsx"

## **M1 – Lectura**

In [ ]:
df=pd.read_excel(ARCHIVO)

print(df.head())

print(df.shape)

     nombre_del_punto_de_monitoreo   latitud   longitud  elevaci_n_m_s_n_m  \
0  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761               2576   
1  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761               2576   
2  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761               2576   
3  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761               2576   
4  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761               2576   

  corriente zona_hidrogr_fica_zh  szh_c_digo_rea_zona_subzona  \
0    BOGOTA       Alto Magdalena                         2120   
1    BOGOTA       Alto Magdalena                         2120   
2    BOGOTA       Alto Magdalena                         2120   
3    BOGOTA       Alto Magdalena                         2120   
4    BOGOTA       Alto Magdalena                         2120   

  nombre_subzona_hidrogr_fica  departamento    municipio  \
0                  Río Bogotá  CUNDINAMARCA  BOGOTÁ D.C.   
1                  R

## **M2 – Validación estructural**

In [ ]:
print(df.dtypes)

print()

print(type(df.loc[0,"fecha"])) # Changed 'Fecha' to 'fecha'

print()

print(df["fecha"].head())

df['fecha'] = pd.to_datetime(df['fecha']) # Convert 'fecha' column to datetime type

print('\nDespués de la conversión a datetime:')
print(df['fecha'].head())
print(df['fecha'].dtype)

nombre_del_punto_de_monitoreo     object
latitud                          float64
longitud                         float64
elevaci_n_m_s_n_m                  int64
corriente                         object
zona_hidrogr_fica_zh              object
szh_c_digo_rea_zona_subzona        int64
nombre_subzona_hidrogr_fica       object
departamento                      object
municipio                         object
fecha                             object
propiedad_observada               object
resultado                         object
unidad_del_resultado              object
proyecto                          object
codigo__muestra                    int64
dtype: object

<class 'str'>

0    2019-12-13T00:00:00.000
1    2019-12-13T00:00:00.000
2    2019-12-13T00:00:00.000
3    2019-12-13T00:00:00.000
4    2019-12-13T00:00:00.000
Name: fecha, dtype: object

Después de la conversión a datetime:
0   2019-12-13
1   2019-12-13
2   2019-12-13
3   2019-12-13
4   2019-12-13
Name: fecha, dtype: datetime6

In [ ]:
print("\nCobertura temporal")
print("-"*30)

print("Inicio :", df["fecha"].min())
print("Fin    :", df["fecha"].max())

print()

print("Fechas inválidas:",
      df["fecha"].isnull().sum())


Cobertura temporal
------------------------------
Inicio : 2005-02-15 00:00:00
Fin    : 2024-06-19 00:00:00

Fechas inválidas: 0


In [ ]:
print(df.dtypes)

print()

print(type(df.loc[0,"fecha"]))

print()

print(df["fecha"].head())

df["fecha"] = pd.to_datetime(
    df["fecha"],
    errors="coerce"
)

print()

print(df["fecha"].dtype)

print()

print("Cobertura temporal")
print("------------------------------")
print("Inicio :", df["fecha"].min())
print("Fin    :", df["fecha"].max())

print()

print("Fechas inválidas:",
      df["fecha"].isnull().sum())

nombre_del_punto_de_monitoreo            object
latitud                                 float64
longitud                                float64
elevaci_n_m_s_n_m                         int64
corriente                                object
zona_hidrogr_fica_zh                     object
szh_c_digo_rea_zona_subzona               int64
nombre_subzona_hidrogr_fica              object
departamento                             object
municipio                                object
fecha                            datetime64[ns]
propiedad_observada                      object
resultado                                object
unidad_del_resultado                     object
proyecto                                 object
codigo__muestra                           int64
dtype: object

<class 'pandas._libs.tslibs.timestamps.Timestamp'>

0   2019-12-13
1   2019-12-13
2   2019-12-13
3   2019-12-13
4   2019-12-13
Name: fecha, dtype: datetime64[ns]

datetime64[ns]

Cobertura temporal
-------------------

In [ ]:
print(df.shape)

print(df["propiedad_observada"].nunique())

print(df["propiedad_observada"].value_counts())

(10875, 16)
68
propiedad_observada
CONDUCTIVIDAD ELECTRICA                   474
pH                                        474
TEMPERATURA                               474
SOLIDOS SUSPENDIDOS TOTALES               473
OXIGENO DISUELTO (OD)                     471
                                         ... 
FENOLES                                     2
HIDROCARBUROS                               2
ESCHERICHIA COLI POR SUSTRATO DEFINIDO      2
GRASAS Y ACEITES                            1
CLORURO                                     1
Name: count, Length: 68, dtype: int64


## **M3 – Estandarización**

In [ ]:
# Crear identificadores

df.insert(
    0,
    "ID_Registro",
    [
        f"{CODIGO_CAPA}-{str(i).zfill(7)}"
        for i in range(1, len(df)+1)
    ]
)

df["ID_Capa"] = CODIGO_CAPA

In [ ]:
# Ordenar columnas

columnas_control = [
    "ID_Registro",
    "ID_Capa"
]

otras_columnas = [
    c for c in df.columns
    if c not in columnas_control
]

df = df[columnas_control + otras_columnas]

In [ ]:
# Normalización de tipos

df["resultado"] = pd.to_numeric(
    df["resultado"],
    errors="coerce"
)

df["fecha"] = pd.to_datetime(
    df["fecha"],
    errors="coerce"
)

df["codigo__muestra"] = df["codigo__muestra"].astype(str)

df["codigoestacion"] = (
    df["nombre_del_punto_de_monitoreo"]
    .astype(str)
)

¿Por qué crear codigoestacion?

Porque las demás capas poseen un identificador principal del punto de monitoreo.

En Calidad de Agua no existe uno.

El campo más estable es: nombre_del_punto_de_monitoreo

Por eso podemos construir un identificador lógico.

Más adelante incluso podremos reemplazarlo por un código hash si lo deseas.

In [ ]:
# Eliminar espacios

df.columns = df.columns.str.strip()

columnas_texto = [
    "nombre_del_punto_de_monitoreo",
    "corriente",
    "zona_hidrogr_fica_zh",
    "nombre_subzona_hidrogr_fica",
    "departamento",
    "municipio",
    "propiedad_observada",
    "unidad_del_resultado",
    "proyecto"
]

for c in columnas_texto:
    df[c] = df[c].astype(str).str.strip()

In [ ]:
# Ordenar registros

df = df.sort_values(
    by=[
        "nombre_del_punto_de_monitoreo",
        "fecha",
        "propiedad_observada"
    ]
).reset_index(drop=True)

In [ ]:
print(df.head())

print()

print(df.columns)

print()

print(df.dtypes)

   ID_Registro ID_Capa    nombre_del_punto_de_monitoreo   latitud   longitud  \
0  CLD-0000001     CLD  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761   
1  CLD-0000002     CLD  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761   
2  CLD-0000003     CLD  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761   
3  CLD-0000004     CLD  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761   
4  CLD-0000005     CLD  BOGOTA_CUN_BOGOTA_CORTIJO EL 37  4.729092 -74.126761   

   elevaci_n_m_s_n_m corriente zona_hidrogr_fica_zh  \
0               2576    BOGOTA       Alto Magdalena   
1               2576    BOGOTA       Alto Magdalena   
2               2576    BOGOTA       Alto Magdalena   
3               2576    BOGOTA       Alto Magdalena   
4               2576    BOGOTA       Alto Magdalena   

   szh_c_digo_rea_zona_subzona nombre_subzona_hidrogr_fica  departamento  \
0                         2120                  Río Bogotá  CUNDINAMARCA   
1                         2120      

In [ ]:
# # Seleccionar los 10 parámetros con mayor frecuencia
#
# top_10 = (
#     df["propiedad_observada"]
#       .value_counts()
#       .nlargest(10)
#       .index
# )
#
# df = df[
#     df["propiedad_observada"].isin(top_10)
# ].copy()
#
# print("Parámetros seleccionados:")
#
# print(top_10.tolist())
#
# print()
#
# print(df.shape)

¿Dónde pondremos el Top 10?

Más adelante.

Cuando lleguemos a la integración de capas (Clima + Hidrología + Calidad + ...)

## **M4 – Gobierno del Dato**

In [ ]:
# Fecha ETL

fecha = datetime.now()

df["Fecha_ETL"] = fecha.strftime("%Y-%m-%d")

df["Hora_ETL"] = fecha.strftime("%H:%M:%S")

In [ ]:
# Versión Framework

df["Version_Framework"] = VERSION

In [ ]:
# Estado del registro

df["Estado_Registro"] = "VALIDADO"

In [ ]:
# Responsable

df["Responsable"] = RESPONSABLE

In [ ]:
# Hash de Integridad

def generar_hash(fila):

    texto = "".join(fila.astype(str))

    return hashlib.md5(texto.encode()).hexdigest()

df["Hash"] = df.astype(str).apply(generar_hash, axis=1)

In [ ]:
df.head()

,ID_Registro,ID_Capa,nombre_del_punto_de_monitoreo,latitud,longitud,elevaci_n_m_s_n_m,corriente,zona_hidrogr_fica_zh,szh_c_digo_rea_zona_subzona,nombre_subzona_hidrogr_fica,...,unidad_del_resultado,proyecto,codigo__muestra,codigoestacion,Fecha_ETL,Hora_ETL,Version_Framework,Estado_Registro,Responsable,Hash
0,CLD-0000001,CLD,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,4.729092,-74.126761,2576,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,µS/cm,PND_2018-2022,27774,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,412a236660e9429fd31d0257f61d152b
1,CLD-0000002,CLD,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,4.729092,-74.126761,2576,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg O2/L,PND_2018-2022,27774,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,3812c64353c2fe1d3c9bdf80c5d1a00c
2,CLD-0000003,CLD,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,4.729092,-74.126761,2576,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg P/L,PND_2018-2022,27774,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,6102b4a76674a384f92e2c896744040f
3,CLD-0000004,CLD,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,4.729092,-74.126761,2576,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg/L,PND_2018-2022,27774,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,1a656252bd4e993ff45b90ab4c3fc57d
4,CLD-0000005,CLD,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,4.729092,-74.126761,2576,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg O2/L,PND_2018-2022,27774,BOGOTA_CUN_BOGOTA_CORTIJO EL 37,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,583ee3434df11e4d93fb5ffec3c850cf


In [ ]:
print(df.columns)

Index(['ID_Registro', 'ID_Capa', 'nombre_del_punto_de_monitoreo', 'latitud',
       'longitud', 'elevaci_n_m_s_n_m', 'corriente', 'zona_hidrogr_fica_zh',
       'szh_c_digo_rea_zona_subzona', 'nombre_subzona_hidrogr_fica',
       'departamento', 'municipio', 'fecha', 'propiedad_observada',
       'resultado', 'unidad_del_resultado', 'proyecto', 'codigo__muestra',
       'codigoestacion', 'Fecha_ETL', 'Hora_ETL', 'Version_Framework',
       'Estado_Registro', 'Responsable', 'Hash'],
      dtype='object')


In [ ]:
print(df.dtypes)

ID_Registro                              object
ID_Capa                                  object
nombre_del_punto_de_monitoreo            object
latitud                                 float64
longitud                                float64
elevaci_n_m_s_n_m                         int64
corriente                                object
zona_hidrogr_fica_zh                     object
szh_c_digo_rea_zona_subzona               int64
nombre_subzona_hidrogr_fica              object
departamento                             object
municipio                                object
fecha                            datetime64[ns]
propiedad_observada                      object
resultado                               float64
unidad_del_resultado                     object
proyecto                                 object
codigo__muestra                          object
codigoestacion                           object
Fecha_ETL                                object
Hora_ETL                                

Observación importante para esta capa

Aquí hay una diferencia importante respecto a las dos primeras capas.

En la Capa Climática y la Capa Hidrológica, cada fila representaba una única observación (clima mensual o nivel hidrológico), por lo que el hash era completamente estable.

En Calidad del Agua, cada fila representa una medición de un parámetro específico (pH, DQO, OD, Conductividad, etc.) realizada en un punto y una fecha determinados. Por ello, el hash se calcula sobre un conjunto de atributos diferente, pero el mecanismo es exactamente el mismo. Esto mantiene la trazabilidad e integridad del dato sin necesidad de modificar el algoritmo.

## **M5 – Metadata**

In [ ]:
metadata = pd.DataFrame({

    "Campo": df.columns,

    "Tipo_Dato": [str(t) for t in df.dtypes],

    "Valores_No_Nulos": df.notnull().sum().values,

    "Valores_Nulos": df.isnull().sum().values,

    "Valores_Unicos": df.nunique().values

})

metadata

,Campo,Tipo_Dato,Valores_No_Nulos,Valores_Nulos,Valores_Unicos
0,ID_Registro,object,10875,0,10875
1,ID_Capa,object,10875,0,1
2,nombre_del_punto_de_monitoreo,object,10875,0,11
3,latitud,float64,10875,0,11
4,longitud,float64,10875,0,11
5,elevaci_n_m_s_n_m,int64,10875,0,11
6,corriente,object,10875,0,1
7,zona_hidrogr_fica_zh,object,10875,0,1
8,szh_c_digo_rea_zona_subzona,int64,10875,0,1
9,nombre_subzona_hidrogr_fica,object,10875,0,1


In [ ]:
descripcion = {

    "ID_Registro":"Identificador único del registro",

    "ID_Capa":"Código de la capa del Framework",

    "nombre_del_punto_de_monitoreo":"Nombre de la estación o punto de monitoreo",

    "latitud":"Latitud geográfica",

    "longitud":"Longitud geográfica",

    "elevaci_n_m_s_n_m":"Elevación sobre el nivel del mar",

    "corriente":"Corriente hídrica monitoreada",

    "zona_hidrogr_fica_zh":"Zona hidrográfica",

    "szh_c_digo_rea_zona_subzona":"Código de la subzona hidrográfica",

    "nombre_subzona_hidrogr_fica":"Nombre de la subzona hidrográfica",

    "departamento":"Departamento",

    "municipio":"Municipio",

    "fecha":"Fecha de la observación",

    "propiedad_observada":"Parámetro de calidad del agua",

    "resultado":"Valor medido",

    "unidad_del_resultado":"Unidad de medida",

    "proyecto":"Proyecto de monitoreo",

    "codigo__muestra":"Código de la muestra",

    "Fecha_ETL":"Fecha del proceso ETL",

    "Hora_ETL":"Hora del proceso ETL",

    "Version_Framework":"Versión del Framework",

    "Estado_Registro":"Estado del registro",

    "Responsable":"Responsable del procesamiento",

    "Hash":"Hash de integridad"

}

metadata["Descripcion"] = metadata["Campo"].map(descripcion)

metadata

,Campo,Tipo_Dato,Valores_No_Nulos,Valores_Nulos,Valores_Unicos,Descripcion
0,ID_Registro,object,10875,0,10875,Identificador único del registro
1,ID_Capa,object,10875,0,1,Código de la capa del Framework
2,nombre_del_punto_de_monitoreo,object,10875,0,11,Nombre de la estación o punto de monitoreo
3,latitud,float64,10875,0,11,Latitud geográfica
4,longitud,float64,10875,0,11,Longitud geográfica
5,elevaci_n_m_s_n_m,int64,10875,0,11,Elevación sobre el nivel del mar
6,corriente,object,10875,0,1,Corriente hídrica monitoreada
7,zona_hidrogr_fica_zh,object,10875,0,1,Zona hidrográfica
8,szh_c_digo_rea_zona_subzona,int64,10875,0,1,Código de la subzona hidrográfica
9,nombre_subzona_hidrogr_fica,object,10875,0,1,Nombre de la subzona hidrográfica


In [ ]:
print(metadata)

                            Campo       Tipo_Dato  Valores_No_Nulos  \
0                     ID_Registro          object             10875   
1                         ID_Capa          object             10875   
2   nombre_del_punto_de_monitoreo          object             10875   
3                         latitud         float64             10875   
4                        longitud         float64             10875   
5               elevaci_n_m_s_n_m           int64             10875   
6                       corriente          object             10875   
7            zona_hidrogr_fica_zh          object             10875   
8     szh_c_digo_rea_zona_subzona           int64             10875   
9     nombre_subzona_hidrogr_fica          object             10875   
10                   departamento          object             10875   
11                      municipio          object             10875   
12                          fecha  datetime64[ns]             10875   
13    

In [ ]:
metadata.to_excel(
    "03_Metadata.xlsx",
    index=False
)

print("Metadata generada correctamente.")

Metadata generada correctamente.


## **M6 – Diccionario de Datos**

In [ ]:
diccionario = pd.DataFrame({

    "Campo":[

        "ID_Registro",
        "ID_Capa",
        "nombre_del_punto_de_monitoreo",
        "latitud",
        "longitud",
        "elevaci_n_m_s_n_m",
        "corriente",
        "zona_hidrogr_fica_zh",
        "szh_c_digo_rea_zona_subzona",
        "nombre_subzona_hidrogr_fica",
        "departamento",
        "municipio",
        "fecha",
        "propiedad_observada",
        "resultado",
        "unidad_del_resultado",
        "proyecto",
        "codigo__muestra",
        "Fecha_ETL",
        "Hora_ETL",
        "Version_Framework",
        "Estado_Registro",
        "Responsable",
        "Hash"

    ]

})

diccionario

,Campo
0,ID_Registro
1,ID_Capa
2,nombre_del_punto_de_monitoreo
3,latitud
4,longitud
5,elevaci_n_m_s_n_m
6,corriente
7,zona_hidrogr_fica_zh
8,szh_c_digo_rea_zona_subzona
9,nombre_subzona_hidrogr_fica


In [ ]:
diccionario["Descripcion"]=[

    "Identificador único del registro",

    "Código de la capa del Framework",

    "Nombre del punto de monitoreo",

    "Latitud geográfica",

    "Longitud geográfica",

    "Elevación sobre el nivel del mar",

    "Corriente hídrica monitoreada",

    "Zona hidrográfica",

    "Código de la subzona hidrográfica",

    "Nombre de la subzona hidrográfica",

    "Departamento",

    "Municipio",

    "Fecha de la medición",

    "Parámetro fisicoquímico o microbiológico",

    "Resultado de la medición",

    "Unidad de medida",

    "Proyecto de monitoreo",

    "Código de la muestra",

    "Fecha del proceso ETL",

    "Hora del proceso ETL",

    "Versión del Framework",

    "Estado del registro",

    "Responsable del procesamiento",

    "Hash de integridad"

]

In [ ]:
diccionario["Tipo_Dato"]=[

    "Texto",
    "Texto",
    "Texto",
    "Decimal",
    "Decimal",
    "Entero",
    "Texto",
    "Texto",
    "Entero",
    "Texto",
    "Texto",
    "Texto",
    "Fecha",
    "Texto",
    "Decimal",
    "Texto",
    "Texto",
    "Entero",
    "Fecha",
    "Hora",
    "Texto",
    "Texto",
    "Texto",
    "Texto"

]

In [ ]:
diccionario["Origen"]=[

    "Framework",
    "Framework",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "CAR",
    "Framework",
    "Framework",
    "Framework",
    "Framework",
    "Framework",
    "Framework"

]

In [ ]:
diccionario["Ejemplo"]=[

    "CLD-0000001",

    "CLD",

    "PUENTE LA VIRGEN",

    "4.6235",

    "-74.1258",

    "2580",

    "RIO BOGOTA",

    "MAGDALENA-CAUCA",

    "2120",

    "RIO BOGOTA",

    "CUNDINAMARCA",

    "CHIA",

    "2019-12-13",

    "OXIGENO DISUELTO (OD)",

    "6.84",

    "mg/L",

    "Monitoreo CAR",

    "1002548",

    "2026-07-14",

    "10:30:55",

    "1.0",

    "VALIDADO",

    "Juan Carlos Riatiga",

    "5ae8e8d7f98a..."

]

In [ ]:
diccionario

,Campo,Descripcion,Tipo_Dato,Origen,Ejemplo
0,ID_Registro,Identificador único del registro,Texto,Framework,CLD-0000001
1,ID_Capa,Código de la capa del Framework,Texto,Framework,CLD
2,nombre_del_punto_de_monitoreo,Nombre del punto de monitoreo,Texto,CAR,PUENTE LA VIRGEN
3,latitud,Latitud geográfica,Decimal,CAR,4.6235
4,longitud,Longitud geográfica,Decimal,CAR,-74.1258
5,elevaci_n_m_s_n_m,Elevación sobre el nivel del mar,Entero,CAR,2580
6,corriente,Corriente hídrica monitoreada,Texto,CAR,RIO BOGOTA
7,zona_hidrogr_fica_zh,Zona hidrográfica,Texto,CAR,MAGDALENA-CAUCA
8,szh_c_digo_rea_zona_subzona,Código de la subzona hidrográfica,Entero,CAR,2120
9,nombre_subzona_hidrogr_fica,Nombre de la subzona hidrográfica,Texto,CAR,RIO BOGOTA


In [ ]:
diccionario.to_excel(

    "04_Diccionario_Datos.xlsx",

    index=False

)

print("Diccionario generado correctamente.")

Diccionario generado correctamente.


## **M7 – Auditoría**

In [ ]:
print("="*70)
print("AUDITORÍA DE LA CAPA")
print("="*70)

print("Registros :",len(df))

print("Variables :",len(df.columns))

print("Columnas")

print(df.columns.tolist())

AUDITORÍA DE LA CAPA
Registros : 10875
Variables : 25
Columnas
['ID_Registro', 'ID_Capa', 'nombre_del_punto_de_monitoreo', 'latitud', 'longitud', 'elevaci_n_m_s_n_m', 'corriente', 'zona_hidrogr_fica_zh', 'szh_c_digo_rea_zona_subzona', 'nombre_subzona_hidrogr_fica', 'departamento', 'municipio', 'fecha', 'propiedad_observada', 'resultado', 'unidad_del_resultado', 'proyecto', 'codigo__muestra', 'codigoestacion', 'Fecha_ETL', 'Hora_ETL', 'Version_Framework', 'Estado_Registro', 'Responsable', 'Hash']


In [ ]:
print("="*70)
print("VALORES NULOS")
print("="*70)

print(df.isnull().sum())

VALORES NULOS
ID_Registro                         0
ID_Capa                             0
nombre_del_punto_de_monitoreo       0
latitud                             0
longitud                            0
elevaci_n_m_s_n_m                   0
corriente                           0
zona_hidrogr_fica_zh                0
szh_c_digo_rea_zona_subzona         0
nombre_subzona_hidrogr_fica         0
departamento                        0
municipio                           0
fecha                               0
propiedad_observada                 0
resultado                        2639
unidad_del_resultado                0
proyecto                            0
codigo__muestra                     0
codigoestacion                      0
Fecha_ETL                           0
Hora_ETL                            0
Version_Framework                   0
Estado_Registro                     0
Responsable                         0
Hash                                0
dtype: int64


In [ ]:
print("="*70)
print("REGISTROS DUPLICADOS")
print("="*70)

print(df.duplicated().sum())

REGISTROS DUPLICADOS
0


In [ ]:
print("="*70)
print("COBERTURA TEMPORAL")
print("="*70)

print("Inicio :",df["fecha"].min())

print("Fin    :",df["fecha"].max())

COBERTURA TEMPORAL
Inicio : 2005-02-15 00:00:00
Fin    : 2024-06-19 00:00:00


In [ ]:
print("="*70)
print("COBERTURA ESPACIAL")
print("="*70)

print("Puntos de monitoreo :",
      df["nombre_del_punto_de_monitoreo"].nunique())

print()

print(df["nombre_del_punto_de_monitoreo"].value_counts())

COBERTURA ESPACIAL
Puntos de monitoreo : 11

nombre_del_punto_de_monitoreo
RCA_BOGOTA_CUN_TOCAIMA_PTE.PORTILLO [21207960]               1441
RCA_BOGOTA_CUN_SOACHA_ALICACHIN - EL SALTO [21207120]        1385
RCA_BOGOTA_CUN_GIRARDOT_CAMPINA LA [21209200]                1385
RCA_BOGOTA_CUN_COTA_PTE.LA VIRGEN [21208110]                 1378
RCA_BOGOTA_CUN_VILLAPINZON_SANPEDRO-BOGOTA                   1209
RCA_BOGOTA_CUN_TOCANCIPA_PTE.TULIO BOTERO-BOGOTA             1204
RCA_BOGOTA_CUN_VILLAPINZON_PTE.CARRETERA-BOGOTA              1157
RCA_BOGOTA_CUN_EL COLEGIO_PTE.CARRETERA LA MESA - MESITAS    1156
BOGOTA_CUN_CHIA_PTE.EL COMUN                                  344
BOGOTA_CUN_BOGOTA_CORTIJO EL 37                               112
BOGOTA_CUN_CHIA_FRENTE HACIENDA DARIEN                        104
Name: count, dtype: int64


In [ ]:
print("="*70)
print("MUNICIPIOS")
print("="*70)

print(df["municipio"].value_counts())

MUNICIPIOS
municipio
VILLAPINZÓN    2366
TOCAIMA        1441
GIRARDOT       1385
SOACHA         1385
COTA           1378
TOCANCIPÁ      1204
EL COLEGIO     1156
CHÍA            448
BOGOTÁ D.C.     112
Name: count, dtype: int64


In [ ]:
print("="*70)
print("PARÁMETROS DE CALIDAD")
print("="*70)

print("Número de parámetros:")

print(df["propiedad_observada"].nunique())

print()

print(df["propiedad_observada"].value_counts())

PARÁMETROS DE CALIDAD
Número de parámetros:
68

propiedad_observada
CONDUCTIVIDAD ELECTRICA                   474
pH                                        474
TEMPERATURA                               474
SOLIDOS SUSPENDIDOS TOTALES               473
OXIGENO DISUELTO (OD)                     471
                                         ... 
FENOLES                                     2
HIDROCARBUROS                               2
ESCHERICHIA COLI POR SUSTRATO DEFINIDO      2
GRASAS Y ACEITES                            1
CLORURO                                     1
Name: count, Length: 68, dtype: int64


In [ ]:
print("="*70)
print("UNIDADES")
print("="*70)

print(df["unidad_del_resultado"].value_counts())

UNIDADES
unidad_del_resultado
mg O2/L          1265
mg/L              967
µS/cm             474
°C                474
mg P-PO4-3/L      419
                 ... 
mg Hg/L             5
µg MALATIÓN/L       5
µg AMETRINA/L       5
NMP/100 cm3         4
mg Cl-/L            1
Name: count, Length: 70, dtype: int64


In [ ]:
print("="*70)
print("INTEGRIDAD HASH")
print("="*70)

print("Hashes únicos:")

print(df["Hash"].nunique())

print()

print("Registros:")

print(len(df))

INTEGRIDAD HASH
Hashes únicos:
10875

Registros:
10875


In [ ]:
auditoria = pd.DataFrame({

    "Indicador":[

        "Registros",

        "Variables",

        "Valores Nulos",

        "Duplicados",

        "Puntos Monitoreo",

        "Municipios",

        "Parámetros",

        "Hashes Únicos"

    ],

    "Valor":[

        len(df),

        len(df.columns),

        df.isnull().sum().sum(),

        df.duplicated().sum(),

        df["nombre_del_punto_de_monitoreo"].nunique(),

        df["municipio"].nunique(),

        df["propiedad_observada"].nunique(),

        df["Hash"].nunique()

    ]

})

auditoria

,Indicador,Valor
0,Registros,10875
1,Variables,25
2,Valores Nulos,2639
3,Duplicados,0
4,Puntos Monitoreo,11
5,Municipios,9
6,Parámetros,68
7,Hashes Únicos,10875


In [ ]:
auditoria.to_excel(

    "05_Auditoria.xlsx",

    index=False

)

print("Auditoría generada correctamente.")

Auditoría generada correctamente.


## **M8 – Indicadores**

In [ ]:
df["unidad_del_resultado"] = (
    df["unidad_del_resultado"]
      .astype(str)
      .str.strip()
      .str.replace("Kg", "kg", regex=False)
      .str.replace("MG/", "mg/", regex=False)
)

In [ ]:
df["unidad_del_resultado"].value_counts()

,count
unidad_del_resultado,
mg O2/L,1265
mg/L,967
µS/cm,474
°C,474
mg P-PO4-3/L,419
NTU,417
mg N-NO3-/L,408
mg N-NH3-/L,401
mg Ni/L,361


In [ ]:
df["unidad_del_resultado"].nunique()

60

Lo que estamos viendo

No tienes un problema de calidad del dataset.

Tienes un problema de heterogeneidad del monitoreo ambiental, que es completamente normal en bases de calidad de agua.

Por ejemplo:

| Parámetro | Registros | Nulos |
| :--- | :---: | :---: |
| Conductividad | 474 | 0 |
| Temperatura | 474 | 0 |
| pH | 474 | 0 |
| Cadmio Total | 345 | 340 |
| Níquel Total | 361 | 343 |
| Plomo Total | 345 | 318 |

¿Qué significa?

No significa que falte el dato.

Significa que en esa campaña ese parámetro no fue analizado.

Es exactamente como publica la CAR sus monitoreos.

In [ ]:
# ============================================================
# M8 – INDICADORES
# ============================================================

print("="*70)
print("INDICADORES DE CALIDAD")
print("="*70)

indicadores = pd.DataFrame({

    "Indicador":[

        "Registros",
        "Variables",
        "Valores Nulos",
        "Duplicados",
        "Cobertura Temporal (años)",
        "Puntos Monitoreo",
        "Municipios",
        "Parámetros",
        "Unidades",
        "Departamentos",
        "Hashes Únicos"

    ],

    "Valor":[

        len(df),

        len(df.columns),

        int(df.isnull().sum().sum()),

        int(df.duplicated().sum()),

        df["fecha"].dt.year.max() - df["fecha"].dt.year.min() + 1,

        df["nombre_del_punto_de_monitoreo"].nunique(),

        df["municipio"].nunique(),

        df["propiedad_observada"].nunique(),

        df["unidad_del_resultado"].nunique(),

        df["departamento"].nunique(),

        df["Hash"].nunique()

    ]

})



INDICADORES DE CALIDAD


In [ ]:
(
    df.isnull()
      .sum()
      .sort_values(ascending=False)
)

,0
resultado,2639
ID_Capa,0
nombre_del_punto_de_monitoreo,0
latitud,0
longitud,0
elevaci_n_m_s_n_m,0
corriente,0
zona_hidrogr_fica_zh,0
ID_Registro,0
szh_c_digo_rea_zona_subzona,0


In [ ]:
df[df["resultado"].isnull()].head(20)

,ID_Registro,ID_Capa,nombre_del_punto_de_monitoreo,latitud,longitud,elevaci_n_m_s_n_m,corriente,zona_hidrogr_fica_zh,szh_c_digo_rea_zona_subzona,nombre_subzona_hidrogr_fica,...,unidad_del_resultado,proyecto,codigo__muestra,codigoestacion,Fecha_ETL,Hora_ETL,Version_Framework,Estado_Registro,Responsable,Hash
216,CLD-0000217,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Cd/L,PND_2018-2022,11718,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,fe200904299b021469d83e01a2fba7c5
217,CLD-0000229,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Cu/L,PND_2018-2022,11718,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,ca53fde9ac3cc2045ca03c82624e9ee5
219,CLD-0000255,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Cr/L,PND_2018-2022,11718,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,5e92773606cd26b46492b21953a11b62
222,CLD-0000294,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Ni/L,PND_2018-2022,11718,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,949ce6adee820bce6cfd6d2433c959d3
227,CLD-0000372,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Pb/L,PND_2018-2022,11718,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,8ee7a95b0b12c140a3a381793138f175
234,CLD-0000218,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Cd/L,PND_2018-2022,12094,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,8216ece36d416c3bcf3ad1896e18d6f1
235,CLD-0000230,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Cu/L,PND_2018-2022,12094,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,4eb9f7f274a19f43fa564942387f6078
237,CLD-0000256,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Cr/L,PND_2018-2022,12094,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,e26ed58060bb42c4f46d8a3d8a8cf33c
240,CLD-0000295,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg Ni/L,PND_2018-2022,12094,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,ca561a02b7e3d5cf4889c66f3ac1ac9f
243,CLD-0000334,CLD,BOGOTA_CUN_CHIA_PTE.EL COMUN,4.863694,-74.029111,2842,BOGOTA,Alto Magdalena,2120,Río Bogotá,...,mg N-NH3-/L,PND_2018-2022,12094,BOGOTA_CUN_CHIA_PTE.EL COMUN,2026-07-15,20:09:37,1.0,VALIDADO,Juan Carlos Riatiga,b3d66e674d0af7d1f891aa0b494581fd


In [ ]:
df[df["resultado"].isnull()][
    [
        "propiedad_observada",
        "unidad_del_resultado"
    ]
].value_counts()

,,count
propiedad_observada,unidad_del_resultado,
NIQUEL TOTAL EN AGUA,mg Ni/L,343
COBRE TOTAL EN AGUA,mg Cu/L,342
CADMIO TOTAL EN AGUA,mg Cd/L,340
PLOMO TOTAL EN AGUA,mg Pb/L,318
CROMO TOTAL EN AGUA,mg Cr/L,249
ZINC TOTAL EN AGUA,mg Zn/L,167
MANGANESO TOTAL EN AGUA,mg Mn/L,116
SULFATO,mg SO4-2/L,73
NITROGENO AMONIACAL,mg N-NH3-/L,73


¿Qué nos está diciendo la auditoría?

Los 2639 valores nulos no están distribuidos aleatoriamente.

Se concentran principalmente en parámetros como:

| Parámetro | Nulos |
| :--- | :---: |
| Níquel total | 343 |
| Cobre total | 342 |
| Cadmio total | 340 |
| Plomo total | 318 |
| Cromo total | 249 |
| Zinc total | 167 |
| Manganeso total | 116 |

Todos ellos corresponden a metales pesados.

Después aparecen otros parámetros menos frecuentes:

* pesticidas
* hidrocarburos
* mercurio
* fenoles
* grasas y aceites
* etc.

Es decir, no es un problema de captura de datos.

¿Qué significa realmente?

La CAR no mide todos los parámetros en todas las campañas.

Por ejemplo:

| Fecha | Parámetro | Resultado |
| :---: | :--- | :---: |
| 15/02/2018 | pH | 7.3 |
| 15/02/2018 | Temperatura | 19.1 |
| 15/02/2018 | Cadmio | No analizado |
| 15/02/2018 | Níquel | No analizado |

Esto es incluso una ventaja

Porque el Framework conserva exactamente la información oficial.

* No está inventando datos.

* No está interpolando.

* No está eliminando registros.

***Eso es justamente lo que uno espera de una capa de adquisición de datos.***


Los valores nulos identificados en la variable resultado corresponden principalmente a parámetros de monitoreo especializado (metales pesados, pesticidas y contaminantes específicos), cuya medición no se realiza en todas las campañas de muestreo. Estos registros fueron conservados íntegramente con el fin de mantener la fidelidad respecto a la fuente oficial y garantizar la trazabilidad de los datos. El tratamiento de dichos valores será abordado únicamente durante la fase de integración y preparación del Dataset Maestro para modelado.

In [ ]:
print(indicadores)

indicadores.to_excel(
    "06_Indicadores.xlsx",
    index=False
)

print("\nIndicadores guardados en '08_Indicadores.xlsx'")

                    Indicador  Valor
0                   Registros  10875
1                   Variables     25
2               Valores Nulos   2639
3                  Duplicados      0
4   Cobertura Temporal (años)     20
5            Puntos Monitoreo     11
6                  Municipios      9
7                  Parámetros     68
8                    Unidades     60
9               Departamentos      1
10              Hashes Únicos  10875

Indicadores guardados en '08_Indicadores.xlsx'


## **M9 – Análisis Exploratorio de Datos (EDA)**

In [ ]:
print("="*70)
print("ESTADÍSTICAS DESCRIPTIVAS")
print("="*70)

print(df.describe(include="all"))

ESTADÍSTICAS DESCRIPTIVAS
        ID_Registro ID_Capa                   nombre_del_punto_de_monitoreo  \
count         10875   10875                                           10875   
unique        10875       1                                              11   
top     CLD-0010806     CLD  RCA_BOGOTA_CUN_TOCAIMA_PTE.PORTILLO [21207960]   
freq              1   10875                                            1441   
mean            NaN     NaN                                             NaN   
min             NaN     NaN                                             NaN   
25%             NaN     NaN                                             NaN   
50%             NaN     NaN                                             NaN   
75%             NaN     NaN                                             NaN   
max             NaN     NaN                                             NaN   
std             NaN     NaN                                             NaN   

             latitud     

Conclusiones del Análisis de la Capa Ambiental

**1. Integridad del Framework**
Todo el gobierno del dato quedó correctamente incorporado, asegurando la trazabilidad del proceso:
* **✓** `ID_Registro`
* **✓** `ID_Capa`
* **✓** `Fecha_ETL`
* **✓** `Hora_ETL`
* **✓** `Responsable`
* **✓** `Hash`
* **✓** `Estado`
* **✓** `Version`

> **Nota de consistencia:** Esto significa que la estructura es completamente consistente con la **Capa Climática** y la **Capa Hidrológica** del proyecto.

---

**2. Cobertura Espacial**
* Se obtuvieron **11 estaciones**, lo cual coincide exactamente con lo esperado para el diseño muestral.
* El filtro por corriente (`BOGOTA`) con **10,875 registros** indica que toda la información pertenece únicamente al **Río Bogotá**, garantizando la delimitación geográfica correcta.

---

**3. Cobertura Temporal**
* **Periodo:** Desde el `2005-02-15` hasta el `2024-06-19`.
* Representa prácticamente **20 años de monitoreo**, lo que se traduce en una serie histórica robusta y de alto valor para análisis de tendencias de largo plazo.

---

**4. Cobertura Temática**
* El dataset cuenta con **60 unidades** de medida y **68 parámetros** evaluados.
* Esta diversidad demuestra la enorme riqueza del conjunto de datos, posicionándola probablemente como la **capa ambiental más compleja del Framework**.

---

**5. Origen del Proyecto**
* Se identifica la participación del **Ideam** junto con otro proyecto institucional.
* **Implicación:** Esto significa que la CAR integró campañas de diferentes proyectos, un dato clave a futuro para identificar y mitigar posibles diferencias metodológicas en los laboratorios o la toma de muestras.

---

**6. Relación de Muestras y Registros**
* Se identificaron **474 muestras** (campañas de muestreo) frente a los **10,875 registros**.
* **Explicación:** Es importante aclarar que no son solo 474 fechas aisladas; son 474 campañas de monitoreo donde cada una mide múltiples parámetros a la vez, lo que explica la densidad del dataset.

---

**7. Análisis de Resultados Vacíos**
* El único aspecto a tener en cuenta es la columna `resultado`, la cual contiene registros sin medición (valores nulos).
* **Conclusión:** Como se analizó previamente, **no es un error de carga**, sino que forma parte de la estructura real del monitoreo (parámetros planificados que no fueron analizados en ciertas campañas).

Mi única recomendación para mejorar el EDA

Agregar una celda adicional (la número 10) para analizar precisamente esos resultados.

In [ ]:
print("="*70)
print("COBERTURA DE MEDICIONES")
print("="*70)

print()

print("Resultados reportados :",
      df["resultado"].notnull().sum())

print("Resultados sin reporte:",
      df["resultado"].isnull().sum())

print()

print(
    "Cobertura:",
    round(
        df["resultado"].notnull().mean()*100,
        2
    ),
    "%"
)

COBERTURA DE MEDICIONES

Resultados reportados : 8236
Resultados sin reporte: 2639

Cobertura: 75.73 %


In [ ]:
print("="*70)
print("VALORES ÚNICOS")
print("="*70)

print(df.nunique())

VALORES ÚNICOS
ID_Registro                      10875
ID_Capa                              1
nombre_del_punto_de_monitoreo       11
latitud                             11
longitud                            11
elevaci_n_m_s_n_m                   11
corriente                            1
zona_hidrogr_fica_zh                 1
szh_c_digo_rea_zona_subzona          1
nombre_subzona_hidrogr_fica          1
departamento                         1
municipio                            9
fecha                              182
propiedad_observada                 68
resultado                         2652
unidad_del_resultado                60
proyecto                             2
codigo__muestra                    474
codigoestacion                      11
Fecha_ETL                            1
Hora_ETL                             1
Version_Framework                    1
Estado_Registro                      1
Responsable                          1
Hash                             10875
dtype: int

In [ ]:
print("="*70)
print("FRECUENCIA DE PARÁMETROS")
print("="*70)

print(
    df["propiedad_observada"]
      .value_counts()
)

FRECUENCIA DE PARÁMETROS
propiedad_observada
CONDUCTIVIDAD ELECTRICA                   474
pH                                        474
TEMPERATURA                               474
SOLIDOS SUSPENDIDOS TOTALES               473
OXIGENO DISUELTO (OD)                     471
                                         ... 
FENOLES                                     2
HIDROCARBUROS                               2
ESCHERICHIA COLI POR SUSTRATO DEFINIDO      2
GRASAS Y ACEITES                            1
CLORURO                                     1
Name: count, Length: 68, dtype: int64


In [ ]:
print("="*70)
print("PUNTOS DE MONITOREO")
print("="*70)

print(
    df["nombre_del_punto_de_monitoreo"]
      .value_counts()
)

PUNTOS DE MONITOREO
nombre_del_punto_de_monitoreo
RCA_BOGOTA_CUN_TOCAIMA_PTE.PORTILLO [21207960]               1441
RCA_BOGOTA_CUN_SOACHA_ALICACHIN - EL SALTO [21207120]        1385
RCA_BOGOTA_CUN_GIRARDOT_CAMPINA LA [21209200]                1385
RCA_BOGOTA_CUN_COTA_PTE.LA VIRGEN [21208110]                 1378
RCA_BOGOTA_CUN_VILLAPINZON_SANPEDRO-BOGOTA                   1209
RCA_BOGOTA_CUN_TOCANCIPA_PTE.TULIO BOTERO-BOGOTA             1204
RCA_BOGOTA_CUN_VILLAPINZON_PTE.CARRETERA-BOGOTA              1157
RCA_BOGOTA_CUN_EL COLEGIO_PTE.CARRETERA LA MESA - MESITAS    1156
BOGOTA_CUN_CHIA_PTE.EL COMUN                                  344
BOGOTA_CUN_BOGOTA_CORTIJO EL 37                               112
BOGOTA_CUN_CHIA_FRENTE HACIENDA DARIEN                        104
Name: count, dtype: int64


In [ ]:
print("="*70)
print("MUNICIPIOS")
print("="*70)

print(
    df["municipio"]
      .value_counts()
)

MUNICIPIOS
municipio
VILLAPINZÓN    2366
TOCAIMA        1441
GIRARDOT       1385
SOACHA         1385
COTA           1378
TOCANCIPÁ      1204
EL COLEGIO     1156
CHÍA            448
BOGOTÁ D.C.     112
Name: count, dtype: int64


In [ ]:
print("="*70)
print("DEPARTAMENTOS")
print("="*70)

print(
    df["departamento"]
      .value_counts()
)

DEPARTAMENTOS
departamento
CUNDINAMARCA    10875
Name: count, dtype: int64


In [ ]:
print("="*70)
print("UNIDADES DE MEDIDA")
print("="*70)

print(
    df["unidad_del_resultado"]
      .value_counts()
)

UNIDADES DE MEDIDA
unidad_del_resultado
mg O2/L                                           1265
mg/L                                               967
µS/cm                                              474
°C                                                 474
mg P-PO4-3/L                                       419
NTU                                                417
mg N-NO3-/L                                        408
mg N-NH3-/L                                        401
mg Ni/L                                            361
mg Zn/L                                            361
mg Cu/L                                            361
mg Cr/L                                            360
mg Cd/L                                            345
mg Pb/L                                            345
mg SO4-2/L                                         332
mg P/L                                             325
Unidades de pH                                     246
mg Fe/L                  

In [ ]:
print("="*70)
print("RANGO TEMPORAL")
print("="*70)

print(df["fecha"].min())

print(df["fecha"].max())

RANGO TEMPORAL
2005-02-15 00:00:00
2024-06-19 00:00:00


In [ ]:
print("="*70)
print("VARIABLES NUMÉRICAS")
print("="*70)

variables = df.select_dtypes(include="number").columns

print(variables.tolist())

VARIABLES NUMÉRICAS
['latitud', 'longitud', 'elevaci_n_m_s_n_m', 'szh_c_digo_rea_zona_subzona', 'resultado']


In [ ]:
print()

print("="*70)
print("EDA GENERADO CORRECTAMENTE")
print("="*70)


EDA GENERADO CORRECTAMENTE


In [ ]:
!pip install xlsxwriter

ruta_salida = "07_EDA.xlsx"

with pd.ExcelWriter(ruta_salida, engine='xlsxwriter') as writer:
    # Estadísticas descriptivas
    df.describe(include='all').to_excel(writer, sheet_name='Descriptive_Statistics')

    # Frecuencia de parámetros
    df['propiedad_observada'].value_counts().to_excel(writer, sheet_name='Parameter_Frequency')

    # Frecuencia de puntos de monitoreo
    df['nombre_del_punto_de_monitoreo'].value_counts().to_excel(writer, sheet_name='Monitoring_Points_Frequency')

    # Frecuencia de municipios
    df['municipio'].value_counts().to_excel(writer, sheet_name='Municipality_Frequency')

    # Frecuencia de departamentos
    df['departamento'].value_counts().to_excel(writer, sheet_name='Department_Frequency')

    # Frecuencia de unidades de medida
    df['unidad_del_resultado'].value_counts().to_excel(writer, sheet_name='Units_Frequency')

print("="*70)
print("REPORTE EDA GENERADO EN '07_EDA.xlsx'")
print("="*70)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 175.3/175.3 kB 6.6 MB/s eta 0:00:00
REPORTE EDA GENERADO EN '07_EDA.xlsx'


## **M10 – README**

In [ ]:
# ============================================================
# M10 – README
# ============================================================

readme = f"""# FRAMEWORK V7

## Capa 3 - Calidad del Agua

### Información General

| Campo | Valor |
|--------|-------|
| Nombre de la capa | {NOMBRE_CAPA} |
| Código | {CODIGO_CAPA} |
| Versión | {VERSION} |
| Responsable | {RESPONSABLE} |
| Fuente | {FUENTE} |

---

## Descripción

Esta capa contiene la información oficial del monitoreo de calidad del agua
de la cuenca del río Bogotá, proveniente de Datos Abiertos Colombia.

La información corresponde a campañas históricas de monitoreo realizadas
por la Corporación Autónoma Regional de Cundinamarca (CAR) y otras entidades,
incluyendo parámetros fisicoquímicos, microbiológicos y de contaminación.

---

## Cobertura

**Cobertura temporal**

- Inicio: {df["fecha"].min()}
- Fin: {df["fecha"].max()}

**Cobertura espacial**

- Puntos de monitoreo: {df["nombre_del_punto_de_monitoreo"].nunique()}
- Municipios: {df["municipio"].nunique()}
- Departamentos: {df["departamento"].nunique()}

---

## Contenido

Registros:

{len(df):,}

Variables:

{len(df.columns)}

Parámetros monitoreados:

{df["propiedad_observada"].nunique()}

Unidades de medida:

{df["unidad_del_resultado"].nunique()}

---

## Gobierno del Dato

La capa incorpora:

- ID_Registro
- ID_Capa
- Fecha_ETL
- Hora_ETL
- Responsable
- Estado_Registro
- Hash MD5

---

## Archivos generados

- 03_Capa_Calidad_Agua_Framework.xlsx
- Diccionario_Capa_Calidad_Agua.xlsx
- Auditoria_Capa_Calidad_Agua.xlsx
- Indicadores_Capa_Calidad_Agua.xlsx
- 07_README.md

---

## Framework

Framework V7 para la aplicación de Tecnologías 4.0 y Ciencia de Datos
en la Gestión Hídrica.

Autor:
Juan Carlos Riatiga

"""

with open("08_README.md", "w", encoding="utf-8") as f:
    f.write(readme)

print("="*70)
print("README GENERADO")
print("="*70)
print()
print("Archivo:")
print("08_README.md")

README GENERADO

Archivo:
08_README.md


## **Exportación Final**

In [ ]:
# ============================================================
# EXPORTACIÓN FINAL
# FRAMEWORK V7
# CAPA 3 - CALIDAD DEL AGUA
# ============================================================

NOMBRE_ARCHIVO = "02_Capa_Calidad_Agua_Framework.xlsx"

# Orden final de columnas

columnas = [

    "ID_Registro",
    "ID_Capa",

    "nombre_del_punto_de_monitoreo",
    "codigoestacion",

    "latitud",
    "longitud",
    "elevaci_n_m_s_n_m",

    "corriente",
    "zona_hidrogr_fica_zh",
    "szh_c_digo_rea_zona_subzona",
    "nombre_subzona_hidrogr_fica",

    "departamento",
    "municipio",

    "fecha",

    "propiedad_observada",
    "resultado",
    "unidad_del_resultado",

    "proyecto",
    "codigo__muestra",

    "Fecha_ETL",
    "Hora_ETL",
    "Version_Framework",
    "Estado_Registro",
    "Responsable",
    "Hash"
]

df = df[columnas]

df.to_excel(
    NOMBRE_ARCHIVO,
    index=False
)

print("="*70)
print("FRAMEWORK V7")
print("CAPA 3 - CALIDAD DEL AGUA")
print("="*70)
print()
print("Archivo generado:")
print()
print(NOMBRE_ARCHIVO)
print()
print("Registros :", len(df))
print("Variables :", len(df.columns))
print()
print("Proceso finalizado correctamente.")